ipapython/ipaldap.py

```diff
@@ -720,18 +721,121 @@ def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
+            ldap.set_option(ldap.OPT_X_TLS_CACERTFILE, cacert)
+
         self.ldap_uri = ldap_uri
         self._start_tls = start_tls
         self._force_schema_updates = force_schema_updates
         self._no_schema = no_schema
         self._decode_attrs = decode_attrs
+        self.cacert = cacert
 
         self.log = log_mgr.get_logger(self)
         self._has_schema = False
         self._schema = None
 
         self._conn = self._connect()
 
+        with self.error_handler():
+            if demand_cert:
+                ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, True)
+                self.conn.set_option(ldap.OPT_X_TLS_DEMAND, True)
+
+            if sasl_nocanon:
+                self.conn.set_option(ldap.OPT_X_SASL_NOCANON, ldap.OPT_ON)
+
+            if start_tls:
+                self.conn.start_tls_s()
+
+    @staticmethod
+    def IPAdmin_init(host='', port=389, cacert=None, debug=None,
+                     ldapi=False, realm=None, protocol=None,
+                     force_schema_updates=True, start_tls=False, ldap_uri=None,
+                     no_schema=False, decode_attrs=True, sasl_nocanon=False,
+                     demand_cert=False):
+        def get_ldap_uri(protocol):
+            if protocol == 'ldaps':
+                return 'ldaps://%s' % format_netloc(host, port)
+            elif protocol == 'ldapi':
+                return 'ldapi://%%2fvar%%2frun%%2fslapd-%s.socket' % (
+                    "-".join(realm.split(".")))
+            elif protocol == 'ldap':
+                return 'ldap://%s' % format_netloc(host, port)
+            else:
+                raise ValueError('Protocol %r not supported' % protocol)
+
+        def guess_protocol():
+            """Return the protocol to use based on flags passed to the constructor
+
+            Only used when "protocol" is not specified explicitly.
+
+            If a CA certificate is provided then it is assumed that we are
+            doing SSL client authentication with proxy auth.
+
+            If a CA certificate is not present then it is assumed that we are
+            using a forwarded kerberos ticket for SASL auth. SASL provides
+            its own encryption.
+            """
+            if cacert is not None:
+                return 'ldaps'
+            elif ldapi:
+                return 'ldapi'
+            else:
+                return 'ldap'
+
+        if not ldap_uri:
+            ldap_uri = get_ldap_uri(protocol or guess_protocol())
+
+        ldap_client = LDAPClient(
+            ldap_uri,
+            force_schema_updates=force_schema_updates,
+            no_schema=no_schema,
+            decode_attrs=decode_attrs,
+            cacert=cacert,
+            sasl_nocanon=sasl_nocanon,
+            demand_cert=demand_cert)
+
+        ldap_client.port = port
+        ldap_client.host = host
+        ldap_client.cacert = cacert
+        ldap_client.ldapi = ldapi
+        ldap_client.realm = realm
+        ldap_client.suffixes = {}
```

Comment:
Are all of these really necessary? `git grep` shows me that `.cacert`, `.realm` and `.suffixes` are not used anywhere and `.ldapi` is used only in `ipa-httpd-kdcproxy`. That leaves us with `.host` and `.port`, which should preferrably be transformed into properties of `LDAPClient` rather than set here.